### PR TITLE
Command Log

### PR DESCRIPTION
**Pull request by psionic-k**
tl;dr This package forked command-log-mode.  Never decided how to add it to MELPA because I changed the feature name and broke the API too hard.

The original command-log-mode did not properly declare a minor mode and had a bunch of other issues.  This version incorporated the open PR's and then attempted to get to a healthier level.  At the least, I'll respond to PR's which I don't expect many of.

However, I don't intend to expand this package because after doing this work, I found keycast.  At some point I confused keycast.el with some Mac program that I obviously couldn't use on Linux.  I added a link or mention of Keycast in the README and package description before writing this recipe.

The original package still shows up in google results, and it's kind of a garbage package, but I changed the feature name, so unless we also get rid of or re-designate the old command-log-mode from MELPA, they are independent packages :crying_cat_face: 

### Brief summary of what the package does

`command-log-toggle` shows the command log in a side window

### Direct link to the package repository

https://github.com/positron-solutions/command-log

### Your association with the package

I forked the MELPA package command-log-mode here:  https://github.com/lewang/command-log-mode

### Relevant communications with the upstream package maintainer

None needed

### Checklist

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

### CODE REVIEW

**Comment by tarsius** [comment]
> However...

Just when I started wondering... :grin:

I've compiled a list of similar packages at one point, see https://github.com/tarsius/keycast/wiki#similar-packages.

We should probably all add a note about similar packages to our READMEs and library summaries, even if it's just a link to the above list.

/cc @chuntaro @lewang [I couldn't find github accounts for Drew and Marco.]

And maybe @lewang would be willing to transfer the packages to @psionic-k? After all, their implementation also started out as a fork of someone else's implementation, in order to address some of its issues.

We should also stop adding new implementation to the pool from which potential users either select randomly or after investing all the alternatives to find the one that works for them.

**Comment by psionic-k** [comment]
> stop adding new implementation

Yep.  If I guilt trip myself, I can find the 3min it takes to add deprecation of the old commands so at least it doesn't break rudely. 

As soon as I started replying, I tried out keypression and keycast and realized there was a quick fix for them all to show whatever commands were launched from M-x.

Guess I can't avoid working on things after all.  I'll be at least upstreaming some fixes to keypression.  It flickers and is all-white out of the box.

**Comment by tarsius** [comment]
> > stop adding new implementation
> 
> Yep. If I guilt trip myself,

:yum:

> As soon as I started replying, I tried out keypression and keycast and realized there was a quick fix for them all to show whatever commands were launched from M-x.

Which in the case of keycast, would be what?

**Comment by psionic-k** [comment]
Instead of the pre-command hook, it was valuable to see the post-command hook to reveal commands issued during completion and to properly see the result of M-x command issuance.  For screen casting this is more informative than `ivy-done` etc.  I'm going to implement watching both pre and post command and allow the user to show either or both for debugging or presentation.

I know lispy likely also shows a difference, sometimes delegating out the interactive call to another less interesting command.

I want to look at the repeat map behavior next.

**Comment by psionic-k** [comment]
https://github.com/positron-solutions/command-log/pull/2  Can ignore CI.  It was complaining about `delete-line` in the linter.

First I did some deprecation.  It wouldn't be terrible if we changed the old package instead of adding a new one.

I also upgrade the capability.  What follows is mostly for @tarsius  who I know loves the command loop.

Sensible M-x feedback is possible, though more special cases likely remain.  Lispy even works almost just right.  I would need to except `lispy-space` I think with some sort of mapping since it doesn't appear to delegate out to `self-insert-command`, which messes up "text" logging a bit.

`counsel-M-x` does not update `this-command` in the `post-command-hook`.

In spite of the wonderful name, `real-this-command' is updated by some packages and not others.  Lispy doesn't seem to change it, but Ivy does.

I could add in posframe support, but it would almost duplicate keypression at that point. I think I can more easily update keypression to use the dual-hook style and incorporate my exceptions.  Keypression also has some simple issues to fix where it fails to specify enough window parameters for example.  I'm fixing it up because it's graphically way better for live demo style presentations when it does work.

I considered doing overlays and sideline, but at lot of the buffers I need to display on are org documents with variable pitch, so it will aesthetically be terrible for live demo style presentation with overlays.

**Comment by psionic-k** [comment]
Should I take over the old command-log-mode slot or go for the new command-log slot?

https://github.com/positron-solutions/command-log/releases/tag/0.2.0

Has support for pre and post-command and exceptions for weirdly behaved pre-commands that don't update `this-command` in the post-command hook.

I'm propagating some of the upgrades into keypression

**Comment by tarsius** [comment]
> Should I take over the old command-log-mode slot or go for the new command-log slot?

I would favor the former, but let's give @lewang more time to join the conversation. Actually, I think it would be a good idea for you to contact him by mail, in case he hasn't seen this conversation, which is likely since he hasn't been very active here the last couple of years.

**Comment by tarsius** [comment]
Make sure to read https://github.com/melpa/melpa/wiki/Unmaintained-Packages-and-Forks before you do that.

**Comment by psionic-k** [comment]
I didn't see any email in a few packages.  Going to update the PR for full takeover of the command-log-mode slot.  I'll just have to update my installation instructions.

Keypression and command-log both have the dual-hook solution available now.  Hope I'm not PR'ing that package next and becoming an industry leader in Emacs command logging.

**Comment by psionic-k** [comment]
Oh should I take both slots since `command-log` is more canonical or just `command-log-mode`?

**Comment by tarsius** [comment]
> Oh should I take both slots since `command-log` is more canonical or just `command-log-mode`?

I would prefer the latter.

**Comment by psionic-k** [comment]
Looks like first element in recipe and the package name is tightly coupled.

If I rename the file to command-log-mode but keep the `command-log` feature, is that the least egregious path?

I don't have the impression that the command-log-mode name was ever a good feature name.

I don't want to stall, but I don't have  strong sense of direction on this issue.

**Comment by tarsius** [comment]
> Looks like first element in recipe and the package name is tightly coupled.

They are not "tightly coupled". They MUST be *identical*.

The package-name symbol in the recipe is completely redundant, one can get the same information from the filename. Because people often get that symbol wrong anyway, e.g., because they copy-pasted a completely unrelated package, I have considered proposing that we stop using that and make the whole list the alist, not just the cdr.

> If I rename the file to command-log-mode but keep the `command-log` feature, is that the least egregious path?

That won't work. `require` similarly requires that the feature name matches the file name.

> I don't have the impression that the command-log-mode name was ever a good feature name.

I was under the impression that (1) you have come to the conclusion that many users interested in such functionality are guided (by search engines etc.) to `command-log-mode` (2) you think there are better options (3) you want to takeover the `command-log-mode` package, to save users from ending up using the "subpar" package as it used to be.

Adding a fork of that package with a better, yet very similar, name, will not achieve that. It will *add* to the confusion of someone who tries to determine what the "best" option is (for their needs).

I have watched (part of) your video. Very nice. Makes me very confident that you would do a great job maintaining such a package.

> I don't want to stall, but I don't have strong sense of direction on this issue.

I think we are still at the "waiting for @lewang to respond" stage. (Have you contacted him by mail? That should help speed things up.)

**Comment by riscy** [comment]
I think we can decipher one of @lewang 's email addresses from this page: https://www.emacswiki.org/emacs/LeWang

**Comment by psionic-k** [comment]
Email done.

In the case we get no reply or "yeah go ahead," then the packaging conversation is relevant.

I think I should switch back to the clm package prefix and feature name.  That's the most correct path I can think of.  I just shouldn't have changed the feature to begin with, but little did I know at that time.

I would keep my old repo up but add a big fat warning to the readme and another warning that would show up at runtime.

**Comment by psionic-k** [comment]
Email done.

In the case we get no reply or "yeah go ahead," then the packaging conversation is relevant.

I think I should switch back to the clm package prefix and feature name.  That's the most correct path I could think of.  I just shouldn't have changed the feature to begin with, but little did I know at that time.

**Comment by riscy** [comment]
Thank you @psionic-k -- let's see if we get a reply.  And btw, I know this is not the most fun process, and I (and the community) owe you gratitude for your willingness to keep the packages we all run maintained.

**Comment by lewang** [comment]
What’s needed from me?

Le

**Comment by tarsius** [comment]
@lewang Very briefly, @psionic-k would like to become the new maintainer of `command-log`. I think he would do a good job, but of course this is up to you.

I won't be available to help with this until the new year. I also hope @psionic-k will respond to you soon. Meanwhile, if you haven't done so yet, I recommend you read what has already been said at https://github.com/melpa/melpa/pull/8813.

Happy holidays!

**Comment by l26wangca** [comment]
I would be happy for someone to take over, I haven't opened this package for years.  Just let me know what you need @psionic-k .

**Comment by psionic-k** [comment]
Everything is good @l26wangca.

This PR will stay blocked on the feature rename I mentioned above, going back to `command-log-mode` and the `clm` prefix.  I'll ping this PR when I have mustered the necessary diligence.

**Comment by psionic-k** [comment]
> transfer

Good points.  

I maintain under the https://github.com/positron-solutions org.

**Comment by riscy** [comment]
Friendly ping just to check how things are going here.

**Comment by psionic-k** [comment]
We can handle this outside of MELPA.  Thanks.

**Comment by psionic-k** [comment]
@l26wangca @lewang  I sent an email, but it's probably an old address.  Ready to start porting over changes from command-log.  Can you go ahead and transfer me the command-log-repo at my organization?  https://github.com/positron-solutions
